pallets/kilt-launch/src/lib.rs

```diff
@@ -0,0 +1,797 @@
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub balance_locks: Vec<(T::AccountId, T::BlockNumber, <T as pallet_balances::Config>::Balance)>,
+		pub transfer_account: T::AccountId,
+		pub vesting: Vec<(T::AccountId, T::BlockNumber, BalanceOf<T>)>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self {
+				balance_locks: Default::default(),
+				transfer_account: Default::default(),
+				vesting: Default::default(),
+			}
+		}
+	}
+
+	// Balance type based on pallet_vesting
+	pub type BalanceOf<T> =
+		<<T as pallet_vesting::Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			// Generate initial custom locking configuration
+			// * who - Account which setting the custom lock for
+			// * length - Number of blocks from  until removal of the lock
+			// * locked - Number of tokens which are locked
+			for (ref who, length, locked) in self.balance_locks.iter() {
+				if !length.is_zero() {
+					let balance = <pallet_balances::Pallet<T>>::free_balance(who);
+					assert!(!balance.is_zero(), "Currencies must be init'd before locking");
+					assert!(
+						balance >= *locked,
+						"Locked balance must not exceed total balance for address {:?}",
+						who.to_string()
+					);
+					assert!(
+						!<BalanceLocks<T>>::contains_key(who),
+						"Account with address {:?} must not occur twice in locking",
+						who.to_string()
+					);
+
+					// Add unlock block to storage
+					<BalanceLocks<T>>::insert(
+						who,
+						LockedBalance::<T> {
+							block: *length,
+							amount: (*locked).saturating_sub(T::UsableBalance::get()),
+						},
+					);
+					// Instead of setting the lock now, we do so in
+					// `migrate_genesis_account`, see there for explanation
+				}
+			}
+
+			// Generate initial vesting configuration, taken from pallet_vesting
+			// * who - Account which we are generating vesting configuration for
+			// * begin - Block when the account will start to vest
+			// * length - Number of blocks from `begin` until fully vested
+			for &(ref who, length, locked) in self.vesting.iter() {
+				if !length.is_zero() {
+					let balance = <<T as pallet_vesting::Config>::Currency as Currency<
+						<T as frame_system::Config>::AccountId,
+					>>::free_balance(who);
+					assert!(!balance.is_zero(), "Currencies must be init'd before vesting");
+					assert!(
+						balance >= locked,
+						"Vested balance must not exceed total balance for address {:?}",
+						who.to_string()
+					);
+					assert!(
+						!<Vesting<T>>::contains_key(who),
+						"Account with address {:?} must not occur twice in vesting",
+						who.to_string()
+					);
+
+					let length_as_balance = T::BlockNumberToBalance::convert(length);
+					let per_block = locked / length_as_balance.max(sp_runtime::traits::One::one());
```

Comment:
Could we use a checked division, just to be sure in case anything changes in the future?

Reply:
It's definitely safer. The code was actually taken from the vesting pallet.

Reply:
It doesn't make sense to do a checked division because we check for `length.is_zero()` a couple lines earlier and even safely do a max check with One.

Reply:
I would not rely on a check performed at the beginning of an if branch, as that might change in future. Furthermore, now that I see it better, I think it would be nicer to have a guard condition replacing the if condition and not indent the whole block within the if branch.
Plus, how is `value.max(1)` protecting against division by zero? Should it be `value.min(1)`?

Reply:
> Plus, how is `value.max(1)` protecting against division by zero? Should it be `value.min(1)`?

That is an error that i make sooooooo often!! Whats the max of 0 and 1?

Reply:
Yes, sorry, the max is not protecting.

Reply:
> > Plus, how is `value.max(1)` protecting against division by zero? Should it be `value.min(1)`?
> 
> That is an error that i make sooooooo often!! Whats the maximum of 0 and 1?

Mhhh, I think we need to move to the irrational numbers to answer correctly 🤔

Reply:
Thanks @weichweich for pointing this out: If `length_as_balance` is zero, then 0.max(1) == 1.

Reply:
`max` is the lower bound, `min` is the upper bound. That's probably where the confusion comes from. 😂

Reply:
Anyway, it kind of proved my point that mistakes, of any kind, can happen 😁 so unless there's some big overhead between checked vs unchecked divisions, I would prefer to use the checked one anything I can.

Reply:
Regarding the guard: I

> Anyway, it kind of proved my point that mistakes, of any kind, can happen 😁 so unless there's some big overhead between checked vs unchecked divisions, I would prefer to use the checked one anywhere I can.

It is overhead and since Parity doesn't do it either, I would not as well: https://github.com/paritytech/substrate/blob/master/frame/vesting/src/lib.rs#L170

Also please note that this function is just the Genesisbuilder, so it won't be called more than once in the lifetime of our chain and the whole pallet will be removed before 2022.

Reply:
Thus, I would leave it as is. The input for this function is a JSON which has one line for each endowed account which might either have
1. All of their balance available
2. All of their balance vested
3. All of their balanced locked

Thus, there are a lot of 0-rows for locking and vesting.

Reply:
I would still say that this is a bit more readable?

```suggestion
					let per_block = locked.checked_div(length_as_balance).unwrap_or(locked);
```

Reply:
Fixed in 6157802e90054e0c5713bf5847c29e924f22226d

Reply:
Fixed guard remark of @Diiaablo95 in ad05d17591d103fa175b65dbbbd71a66a072a36e

> I would not rely on a check performed at the beginning of an if branch, as that might change in future. Furthermore, now that I see it better, I think it would be nicer to have a guard condition replacing the if condition and not indent the whole block within the if branch.